### PR TITLE
Enabled metal on ios simulator

### PR DESCRIPTION
**Pull request by Kavantix**
When apple released ios 13 and catelina it also enabled the ios simulator to use metal.
It only works on catelina and up though.

This PR enabled this feature, however, I also needed to make some changes in skia but not sure where I can commit that.

This is the diff for skia:
```diff
--- a/src/image/SkSurface_GpuMtl.mm
+++ b/src/image/SkSurface_GpuMtl.mm
@@ -27,6 +27,9 @@
 #import <QuartzCore/CAMetalLayer.h>
 #import <MetalKit/MetalKit.h>

+#if TARGET_IPHONE_SIMULATOR
+API_AVAILABLE(ios(13.0))
+#endif  // TARGET_IPHONE_SIMULATOR
 sk_sp<SkSurface> SkSurface::MakeFromCAMetalLayer(GrContext* context,
                                                  GrMTLHandle layer,
                                                  GrSurfaceOrigin origin,
```

### CODE REVIEW

**Comment by chinmaygarde** [comment]
xref https://github.com/flutter/flutter/issues/52358

**Comment by chinmaygarde** [comment]
I have submitted a patch to Skia that should unblock this effort https://skia-review.googlesource.com/c/skia/+/284988

**Comment by Kavantix** [comment]
@chinmaygarde thanks for the quick patch for skia.
I have rebased on master and cleaned up the code a bit.

File: shell/gpu/gpu_surface_metal.h
Comment (chinmaygarde) [inline shell/gpu/gpu_surface_metal.h:21]: Can you use `SK_API_AVAILABLE_CA_METAL_LAYER` here please? There is currently an issue with the compiler version we use where decorating C++ classes with API_AVAILABLE marks the dylib symbol visibility as default instead of hidden. This is why we added the SK_API_AVAILABLE macro is Skia as well. That macro is wrapped by SK_API_AVAILABLE_CA_METAL_LAYER. Just using that macro directly should give you the benefit of not making dylib symobls visible and also the if-else macros.

File: shell/platform/darwin/ios/ios_surface_metal.h
Comment (chinmaygarde) [inline shell/platform/darwin/ios/ios_surface_metal.h:17]: Ditto about `SK_API_AVAILABLE_CA_METAL_LAYER` usage.

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (chinmaygarde) [inline shell/platform/darwin/ios/rendering_api_selection.mm:32]: This is not necessary in Objective-C. If device is `nil`, then sending a `supportsFeatureSet:` message to a `nil` object will return a zero-ed out return value (`NO`).

File: shell/platform/darwin/ios/ios_surface.mm
Comment (chinmaygarde) [inline shell/platform/darwin/ios/ios_surface.mm:42]: Here and elsewhere, please use TARGET_OS_SIMULATOR instead of TARGET_IPHONE_SIMULATOR. The latter is deprecated by Apple in `TargetConditionals.h`.

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (chinmaygarde) [inline shell/platform/darwin/ios/rendering_api_selection.mm:25]: It may be time to put 13.0 in a 10.0 and a `constexpr double` declaration in the header. Maybe something like `constexpr double kMetalOSVersionBaseline = <>` and use that in the @available macros elsewhere. Think of it this way, if we wanted to update the baseline, would modifying just `rendering_api_selection` (header or implementation) be enough? Right now, we would need to modify this translation unit and also hunt for all the places we hardcoded 13.0 or 10.0. Let's isolate them all into one variable that everyone else can refer to.

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (chinmaygarde) [inline shell/platform/darwin/ios/rendering_api_selection.mm:66]: If this is always true but you have to add the @available for the compiler, add an assertion right after this clause that tears down the process. Just `FML_CHECK(false) << "The reason in the comment"` should be sufficient.

Comment (chinmaygarde) [reply shell/platform/darwin/ios/rendering_api_selection.mm:66]: The assertion with comment is generally a good idea anywhere else you had to add a check seemingly unnecessarily. That give future readers a clue about why that was added.

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (Kavantix) [reply shell/platform/darwin/ios/rendering_api_selection.mm:25]: sounds like a plan
Where should I put that constant?

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (Kavantix) [reply shell/platform/darwin/ios/rendering_api_selection.mm:32]: Ah right, has been a while since I wrote any significant amount of objective C

File: shell/platform/darwin/ios/ios_surface.mm
Comment (Kavantix) [reply shell/platform/darwin/ios/ios_surface.mm:42]: If we have the constant which will be 10 or 13 atm
Then the whole if is no longer needed right?

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (Kavantix) [reply shell/platform/darwin/ios/rendering_api_selection.mm]: Using a `constexpr` is not possible since the version number is not actually a double (it can for instance be 13.0.1) so I used a `#define` instead

File: shell/platform/darwin/ios/ios_surface.mm
Comment (chinmaygarde) [inline shell/platform/darwin/ios/ios_surface.mm:52]: Nit: `FML_CHECK(false)`  in an else clause saying this should already have been checked.

File: shell/platform/darwin/ios/rendering_api_selection.h
Comment (chinmaygarde) [inline shell/platform/darwin/ios/rendering_api_selection.h:28]: `METAL_IOS_VERSION_BASELINE` is the convention for macros.

File: shell/platform/darwin/ios/ios_surface.mm
Comment (Kavantix) [reply shell/platform/darwin/ios/ios_surface.mm:52]: That doesnt make sense to me, it not being available is valid since it will use the software renderer in that case

**Comment by Kavantix** [comment]
@chinmaygarde Whats the current roadmap for this?

**Comment by zanderso** [comment]
@chinmaygarde ping

**Comment by xster** [comment]
Looks like Chinmay already approved this. Though @Kavantix if you don't mind, you still have to rebase your commits onto the master branch's tip of tree to resolve the merge conflicts.

**Comment by Kavantix** [comment]
@xster I resolved the conflicts
One of the golden files seems to fail regarding platformviews, could this simply be a difference between the software and metal renderer?

**Comment by dnfield** [comment]
This is almost certainly failing due to differences between software and metal rendering for golden files.

We should do one of the following:

- Enable gold for this (/cc @Piinks) - but I don't think that's trivial
- Update the screenshots for Metal, which might be tricky if the bots are using slightly different GPUs and might have slight differences due to that
- Force software rendering for the tests.

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (dnfield) [inline shell/platform/darwin/ios/rendering_api_selection.mm:43]: We could have something above this if `--enable-software-rendering` has been specified to force the emulator to run using software rendering. I think we should do this for parity with Android, and then for now we could make the bots run using software rendering until we're ready to update goldens.

**Comment by dnfield** [comment]
Left a couple more comments in the code - I think to get this unblocked for now we should respect the `--enable-software-rendering` flag like we do on Android emulator builds, and specify that flag on CI until we're ready to test the Metal backend on a simulator.

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (dnfield) [reply shell/platform/darwin/ios/rendering_api_selection.mm:43]: An easy way to do this would be to add a parameter to `GetRenderingAPIForProcess` about whether to force software rendering, and pass `settings.enable_software_rendering` as the parameter.

If that's set to true, we just return `IOSRenderingAPI::kSoftware`.

File: shell/gpu/gpu_surface_metal.h
Comment (Kavantix) [reply shell/gpu/gpu_surface_metal.h:21]: This was by request of @chinmaygarde

**Comment by Kavantix** [comment]
I do definitely agree that we should respect the force software rendering option.
However, it feels a bit hacky as a solution to get the test green since it would essentially mean that the metal renderer (which everyone will be using by default) will not be tested.

Some questions about the testing:
1. Does the testing stop once it encounters a failed golden file or is this one file the only one that would need changing?
1. Can we somehow get the file that fails which is currently created by the test to visually inspect it?
1. Would using metal for testing potentially speed it up compared to using the software renderer?

Btw, maintainers should be able to push to my branch.

**Comment by dnfield** [comment]
It would be better to just test the goldens on metal, but the problem is that we don't really have a great setup right now to do that. We test on a fleet of mac machines that are not all guaranteed to have the same GPU. Each different model might have slight pixel differences because of differences in GPU or driver.

I would be fine moving this patch forward with just continuing to test using a software backend, and as we get Skia gold enabled for the engine repo we could more easily manage different golden variations for the different GPU(s) in the fleet.

**Comment by dnfield** [comment]
We could still add some test that does _not_ do golden file comparison and is run using the metal renderer on a simulator.

**Comment by dnfield** [comment]
@kavantix would you prefer if someone pushed such changes here? I usually try to avoid doing that unless a patch is very obviously stalled. If you're willing to make these changes that would be good.

**Comment by Kavantix** [comment]
Yes feel free to push any changes here
I don’t have the engine repo setup atm since I needed some extra disk space so although I am willing it might take a while for me to get to this
Especially getting the setting to the correct place would most likely take me quite a bit more time then you guys.

Just have to say I’m glad there is some more attention here now, it was so nice to use the metal backed simulator when I had the repo still so looking forward to using it again!

Btw, would there be any chance of cherry picking this to stable once we get it merged?

**Comment by Kavantix** [comment]
> It would be better to just test the goldens on metal, but the problem is that we don't really have a great setup right now to do that. We test on a fleet of mac machines that are not all guaranteed to have the same GPU. Each different model might have slight pixel differences because of differences in GPU or driver.
> 
> I would be fine moving this patch forward with just continuing to test using a software backend, and as we get Skia gold enabled for the engine repo we could more easily manage different golden variations for the different GPU(s) in the fleet.

Ah thats why thanks for explaining!
And in that case I fully agree with your plan.

**Comment by dnfield** [comment]
I have something that should be working for this, working on a couple tests.

**Comment by dnfield** [comment]
I think this should work, looking for it to run through on CI because goldens are all wrong for me locally.

**Comment by dnfield** [comment]
@chinmaygarde - can you re-review this for my changes?

@cyanglaz FYI and if you have any feedback - this is forcing platform view golden tests to continue to use software rendering. We should look to see if we can enable metal rendering goldens for them. Filed https://github.com/flutter/flutter/issues/67573

**Comment by Piinks** [comment]
> I think this should work, looking for it to run through on CI because goldens are all wrong for me locally.

@dnfield Golden file tests are not executed on CI in flutter/engine. They are executed when it rolls into flutter/flutter. Flutter Gold has to associate images with commits in the flutter/flutter repository, so they can't run from the flutter/engine repository.

File: shell/platform/darwin/ios/rendering_api_selection.h
Comment (chinmaygarde) [inline shell/platform/darwin/ios/rendering_api_selection.h:27]: Please document how these baselines were selected.

**Comment by dnfield** [comment]
@Piinks - the iOS tests have their own golden file logic currently, which doesn't depend on Skia gold - but it means it's more complicated to update and keep up to date for changes in the SDK version(s), different machine OS versions, different GPUs, etc... sorry for confusion.

**Comment by Piinks** [comment]
> @Piinks - the iOS tests have their own golden file logic currently, which doesn't depend on Skia gold - but it means it's more complicated to update and keep up to date for changes in the SDK version(s), different machine OS versions, different GPUs, etc... sorry for confusion.

Ah ok! No worries! I was just checking. ;)

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/RenderingSelectionTest.m
Comment (Kavantix) [inline testing/scenario_app/ios/Scenarios/ScenariosUITests/RenderingSelectionTest.m:29]:
```suggestion
  self.application.launchArguments = @[ @"--animated-color-square", @"--assert-ca-layer-type" ];
```

File: testing/scenario_app/ios/Scenarios/ScenariosUITests/RenderingSelectionTest.m
Comment (Kavantix) [inline testing/scenario_app/ios/Scenarios/ScenariosUITests/RenderingSelectionTest.m:22]:
```suggestion
      @[ @"--animated-color-square", @"--assert-ca-layer-type", @"--enable-software-rendering" ];
```

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (chinmaygarde) [inline shell/platform/darwin/ios/rendering_api_selection.mm:33]: This allows software rendering on devices. This check should only be performed on simulators.

File: shell/platform/darwin/ios/framework/Source/FlutterView.h
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterView.h:38]: Nit: `readwrite` is redundant.

**Review by Kavantix — COMMENTED** [review]
Found a typo, looks good besides that

**Review by chinmaygarde — CHANGES_REQUESTED** [review]
Mostly there. Just need to make sure we don't allow software rendering on devices.

File: shell/platform/darwin/ios/rendering_api_selection.h
Comment (dnfield) [reply shell/platform/darwin/ios/rendering_api_selection.h:27]: This was added by @kavantix - Adding what I assume is correct as 

```cpp
// Metal support was added for simulators in iOS 13.
// Otherwise, the lowest supported version in Skia is iOS 10.
```

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (dnfield) [reply shell/platform/darwin/ios/rendering_api_selection.mm:33]: Done

File: shell/platform/darwin/ios/framework/Source/FlutterView.h
Comment (dnfield) [reply shell/platform/darwin/ios/framework/Source/FlutterView.h:38]: Removed

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (dnfield) [reply shell/platform/darwin/ios/rendering_api_selection.mm:35]: Added an #elif block if we're not in release to print a warning that you can't use this on physical devices.

File: shell/platform/darwin/ios/rendering_api_selection.h
Comment (dnfield) [reply shell/platform/darwin/ios/rendering_api_selection.h:34]: Ahh I see, I'm moving the old comment from the impl file to here, and adding another line about when sim support was added.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:163]: This object is leaked when passed to `[FlutterEngine initWithName:project:allowHeadlessExecution:]` as that ctor retains the `project` arg when not `nil`. Either `autorelease` the object or use a `scoped_nsobject` for the same. Engine TUs don't use ARC.

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (chinmaygarde) [inline shell/platform/darwin/ios/rendering_api_selection.mm:35]: Good call on the log. Why `!FLUTTER_RELEASE` though? This is still a warning log that will only show up when verbose logging is enabled. I'd just use `#else`.

**Review by chinmaygarde — CHANGES_REQUESTED** [review]
Sorry, I noticed a leak as I was reviewing the updates (which look good BTW).

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm
Comment (dnfield) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:163]: Ahh good catch, fixed with autorelease.

File: shell/platform/darwin/ios/rendering_api_selection.mm
Comment (dnfield) [reply shell/platform/darwin/ios/rendering_api_selection.mm:35]: I figure avoid the overhead if you're on release, since you should've seen it already in debug/profile if you're trying. I guess it's probably not a ton of overhead to do the bool check though.. I'll change it to else.

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**
